### PR TITLE
[mesh-forwarder] add time-in-queue statistics collection feature

### PR DESCRIPTION
**Pull request by abtink**
This commit adds the ability to collect statistics on the time messages spend in the TX queue.

When enabled, a histogram of the time-in-queue is collected. The time-in-queue is tracked for direct transmissions only, and is measured as the duration from when a message is added to the TX queue until it is passed to the MAC layer for transmission or dropped.

The histogram data consists of a number of bins, each representing a range of time-in-queue values. The bin interval length and maximum tracked interval can be configured using newly added configuration options.

This commit also adds a new public OT API and a related CLI command (`timeinqueue`) to get the collected statistics or reset the data.

### CODE REVIEW

File: src/core/config/mesh_forwarder.h [hunk 1]
```diff
@@ -148,4 +150,56 @@
...
+ *
+ * The histogram data consists of number of bins, each representing a range of time-in-queue values. The bin interval
+ * length is specified by the `OPENTHREAD_CONFIG_TX_QUEUE_STATISTICS_HISTOGRAM_BIN_INTERVAL` configuration, and the
+ * maximum tracked interval is given by the `OPENTHREAD_CONFIG_TX_QUEUE_STATISTICS_HISTOGRAM_MAX_INTERVAL`.
+ *
+ * Along with histogram, the maximum observed time-in-queue is also tracked.
+ *
+ */
+#ifndef OPENTHREAD_CONFIG_TX_QUEUE_STATISTICS_ENABLE
+#define OPENTHREAD_CONFIG_TX_QUEUE_STATISTICS_ENABLE (OPENTHREAD_CONFIG_BORDER_ROUTING_ENABLE && OPENTHREAD_FTD)
```
Comment (jwhui) [inline src/core/config/mesh_forwarder.h:170]: Do we want `BORDER_ROUTER` or `BORDER_ROUTING` here? :D

File: src/core/config/mesh_forwarder.h [hunk 2]
```diff
@@ -148,4 +150,56 @@
...
+ */
+#ifndef OPENTHREAD_CONFIG_TX_QUEUE_STATISTICS_HISTOGRAM_MAX_INTERVAL
+#define OPENTHREAD_CONFIG_TX_QUEUE_STATISTICS_HISTOGRAM_MAX_INTERVAL \
+    OPENTHREAD_CONFIG_DELAY_AWARE_QUEUE_MANAGEMENT_MARK_ECN_INTERVAL
+#endif
+
+/**
+ * @def OPENTHREAD_CONFIG_TX_QUEUE_STATISTICS_HISTOGRAM_BIN_INTERVAL
+
+ *
```
Comment (jwhui) [inline src/core/config/mesh_forwarder.h:192]:
```suggestion
 * @def OPENTHREAD_CONFIG_TX_QUEUE_STATISTICS_HISTOGRAM_BIN_INTERVAL
 *
```

File: src/core/config/mesh_forwarder.h [hunk 3]
```diff
@@ -148,4 +150,56 @@
...
+ * Specifies the time-in-queue histogram bin interval in milliseconds when the TX queue time-in-queue statistics
+ * collection feature is enabled.
+ *
+ * The number of bins is calculated by dividing `OPENTHREAD_CONFIG_TX_QUEUE_STATISTICS_HISTOGRAM_MAX_INTERVAL` by
+ * `OPENTHREAD_CONFIG_TX_QUEUE_STATISTICS_HISTOGRAM_BIN_INTERVAL` and rounding up to the nearest integer.
+ *
+ */
+#ifndef OPENTHREAD_CONFIG_TX_QUEUE_STATISTICS_HISTOGRAM_BIN_INTERVAL
+
+#define OPENTHREAD_CONFIG_TX_QUEUE_STATISTICS_HISTOGRAM_BIN_INTERVAL 10
```
Comment (jwhui) [inline src/core/config/mesh_forwarder.h:202]:
```suggestion
#ifndef OPENTHREAD_CONFIG_TX_QUEUE_STATISTICS_HISTOGRAM_BIN_INTERVAL
#define OPENTHREAD_CONFIG_TX_QUEUE_STATISTICS_HISTOGRAM_BIN_INTERVAL 10
```

File: src/core/thread/mesh_forwarder.cpp [hunk 4]
```diff
@@ -571,6 +595,9 @@ Message *MeshForwarder::PrepareNextDirectTransmission(void)
         switch (error)
         {
         case kErrorNone:
+#if OPENTHREAD_CONFIG_TX_QUEUE_STATISTICS_ENABLE
+            mTxQueueStats.UpdateFor(*curMessage);
+#endif
```
Comment (jwhui) [inline src/core/thread/mesh_forwarder.cpp:600]: This doesn't capture the time spent actually trying to deliver the message. Should we be recording the time in something like `UpdateSendMessage()`?

Comment (abtink) [reply src/core/thread/mesh_forwarder.cpp:600]: Yes. But that's intentional. The idea is to use collected time-in-queue histogram info to validate/tune parameters for "delay-aware queue management" (which uses this the time message is in queue till it is prepared to be sent).

File: src/core/config/mesh_forwarder.h (hunk 1 above)
Comment (abtink) [reply src/core/config/mesh_forwarder.h]: Changed to `BORDER_ROUTER`.

File: src/cli/cli.cpp [hunk 5]
```diff
@@ -6715,6 +6715,137 @@ template <> otError Interpreter::Process<Cmd("thread")>(Arg aArgs[])
...
+    /**
+     * @cli timeinqueue max
+     * @code
+     * timeinqueue max
+     * 281
+     * Done
+     * @endcode
+     * @par api_copy
+     * #otThreadGetMaxTimeInQueue
+     */
```
Comment (Vyrastas) [inline src/cli/cli.cpp:6822]: Please add a csa tag for the reset command here too 

`@csa{timeinqueue reset}`

**Review by Vyrastas — CHANGES_REQUESTED** [review]

File: include/openthread/thread.h [hunk 6]
```diff
@@ -902,6 +902,60 @@ const otIpCounters *otThreadGetIp6Counters(otInstance *aInstance);
...
+ * Gets the time-in-queue histogram for messages in the TX queue.
+ *
+ * Requires `OPENTHREAD_CONFIG_TX_QUEUE_STATISTICS_ENABLE`.
+ *
+ * Histogram of the time-in-queue of messages in the transmit queue is collected. The time-in-queue is tracked for
+ * direct transmissions only and is measured as the duration from when a message is added to the transmit queue until
+ * it is passed to the MAC layer for transmission or dropped.
+ *
+ * The histogram is returned as an array of `uint32_t` values with `aNumBins` entries. The first entry (at index 0)
+ * represents the number of messages with a time-in-queue in the range `[0, aBinInterval)`. The second entry represents
```
[the remark this answers is not in the document]
Comment (abtink) [reply include/openthread/thread.h]: I intended these as close/open interval notation -> https://en.wikipedia.org/wiki/Interval_(mathematics)

`[` or `]` for closed (include the number in the range, and `(` or `)` for open (exclude the number in the range).

I reworded this comment:
```
 * The histogram is returned as an array of `uint32_t` values with `aNumBins` entries. The first entry in the array
 * (at index 0) represents the number of messages with a time-in-queue less than `aBinInterval`. The second entry
 * represents the number of messages with a time-in-queue greater than or equal to `aBinInterval`, but less than
 * `2 * aBinInterval`. And so on. The last entry represents the number of messages with time-in-queue  greater than or
 * equal to `(aNumBins - 1) * aBinInterval`. 
 ```

File: src/cli/README.md [hunk 7]
```diff
@@ -3277,6 +3278,97 @@ Get the Thread Version number.
...
 
+### timeinqueue
+
+Print the tx queue time-in-queue histogram.
+
+Requires `OPENTHREAD_CONFIG_TX_QUEUE_STATISTICS_ENABLE`.
+
+The time-in-queue is tracked for direct transmissions only and is measured as the duration from when a message is added to the transmit queue until it is passed to the MAC layer for transmission or dropped.
+
+Each table row shows min and max time-in-queue (in milliseconds) followed by number of messages with time-in-queue within the specified min-max range.
```
[the remark this answers is not in the document]
Comment (abtink) [reply src/cli/README.md]: Good idea. Added this in new push (in `README.md` and also in `thread.h` API doc).

File: src/core/thread/mesh_forwarder.hpp [hunk 8]
```diff
@@ -311,6 +311,53 @@ class MeshForwarder : public InstanceLocator, private NonCopyable
...
+    /**
+     * Gets the time-in-queue histogram for messages in the TX queue.
+     *
+     * Histogram of the time-in-queue of messages in the transmit queue is collected. The time-in-queue is tracked for
+     * direct transmissions only and is measured as the duration from when a message is added to the transmit queue
+     * until it is passed to the MAC layer for transmission or dropped.
+     *
+     * The histogram is returned as an array of `uint32_t` values with `aNumBins` entries. The first entry (at index 0)
+     * represents the number of messages with a time-in-queue in the range `[0, aBinInterval)`. The second entry
+     * represents the number of messages with a time-in-queue in the range `[aBinInterval, 2 * aBinInterval)`, and so
```
[the remark this answers is not in the document]
Comment (abtink) [reply src/core/thread/mesh_forwarder.hpp]: Updated in new push.

File: src/cli/README.md [hunk 9]
```diff
@@ -3277,6 +3278,97 @@ Get the Thread Version number.
...
+|  460 |  469 |       0 |
+|  470 |  479 |       0 |
+|  480 |  489 |       0 |
+|  490 |  inf |       0 |
+Done
+```
+
+### timeinqueue max
+
+Print the maximum observed time-in-queue in milliseconds
```
Comment (Vyrastas) [inline src/cli/README.md:3352]: nit: period at the end of the sentence

**Review by Vyrastas — APPROVED** [review]

File: src/cli/README.md (hunk 9 above)
Comment (abtink) [reply src/cli/README.md]: Thanks. Changed in new push (rebase)

**Review by jwhui — APPROVED** [review]

**Pull request merged**